File: assgn2/Deque.java
/*
 * Author:  Param Singh
 * Date:    03/07/2015
 */
import java.util.NoSuchElementException;
import java.util.Iterator;

public class Deque<Item> implements Iterable<Item> {

    private Node first, last;
    private int size;

    private class Node {
        private Item item;
        private Node next;
        private Node prev;
    }

    public Deque() {
       first = null;
       last = null;
    }

    public boolean isEmpty() {
        return first == null;
    }

    public int size() {
        return size;
    }

    public void addFirst(Item item) {

        if (item == null)
            throw new NullPointerException("Can't add null item");

        Node x = new Node();
        x.next = first;
        x.item = item;
        if (first == null)
            last = x;
        else
            first.prev = x;
        first = x;

        size += 1;
    }

    public void addLast(Item item) {

        if (item == null)
            throw new NullPointerException("Can't add null item");

        Node x = new Node();
        x.item = item;
        if (last == null)
            first = x;
        else
            last.next = x;
        x.prev = last;
        last = x;

        size += 1;
    }

    public Item removeFirst() {
        if (isEmpty())
            throw new NoSuchElementException("Can't remove from an empty deque");

        Item val = first.item;

        if (first == last) {
            first.item = null;
            first.next = null;
            first.prev = null;
            first = null;
            last = null;
        }
        else {
            first.item = null;
            Node nxt = first.next;
            first.next = null;
            nxt.prev = null;
            first = nxt;
        }

        size -= 1;

        return val;
    }

    public Item removeLast() {
        if (isEmpty())
            throw new NoSuchElementException("Can't remove from an empty deque");

        Item val = last.item;

        if (first == last) {
            first.item = null;
            first.next = null;
            first.prev = null;
            first = null;
            last = null;
        }
        else {
            last.item = null;
            Node prv = last.prev;
            last.prev = null;
            prv.next = null;
            last = prv;
        }

        size -= 1;

        return val;
    }

    public Iterator<Item> iterator() {
        return new DequeIterator();
    }

    private class DequeIterator implements Iterator<Item> {
        private Node cur;
        public DequeIterator() {
            cur = first;
        }

        public boolean hasNext() {
            return cur != null;
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }

        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();

            Item val = cur.item;
            cur = cur.next;
            return val;
        }
    }
}

File: python/queues/deque.py
class Deque:
    class Node:
        def __init__(self):
            self.item = None
            self.nxt = None
            self.prev = None

    def __init__(self):
        self._first = None
        self._last = None
        self._size = 0

    def is_empty(self):
        return self._size == 0

    def size(self):
        return self._size

    def add_first(self, item):
        # Check if item is valid
        if item == None:
            raise ValueError("Can't add None to deque")
        x = Deque.Node()
        x.item = item
        x.nxt = self._first
        if self.is_empty():
            self._last = x
        else:
            self._first.prev = x
        self._first = x
        self._size += 1

    def add_last(self, item):
        # Check if item is valid
        if item == None:
            raise ValueError("Can't add None to deque")
        x = Deque.Node()
        x.item = item
        if self.is_empty():
            self._first = x
        else:
            self._last.nxt = x
        x.prev = self._last
        self._last = x

        self._size += 1

    def remove_first(self):
        if self.is_empty():
            raise Exception("Can't remove from an empty deque")

        val = self._first.item
        # if there is only one element in deque
        if self._first == self._last:
            self._first = None
            self._last = None
        else:
            new_first = self._first.nxt
            new_first.prev = None
            self._first = new_first

        self._size -= 1
        return val

    def remove_last(self):
        if self.is_empty():
            raise Exception("Can't remove from an empty deque")

        val = self._last.item
        if self._first == self._last:
            self._first = self._last = None
        else:
            new_last = self._last.prev
            new_last.nxt = None
            self._last = new_last

        self._size -= 1
        return val

    class DequeIterator:
        def __init__(self, first):
            self._cur = first

        def next(self):
            if self._cur.nxt:
                val = self._cur.item
                self._cur = self._cur.nxt
                return val
            else:
                raise StopIteration()

    def __iter__(self):
        return Deque.DequeIterator(self._first)
